org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/format/LSPFormatHandler.java

```diff
@@ -50,8 +52,15 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
 			final Shell shell = textEditor.getSite().getShell();
 			ISelection selection = HandlerUtil.getCurrentSelection(event);
 			if (document != null && selection instanceof ITextSelection textSelection) {
+				final long originalStamp = LSPEclipseUtils.getDocumentModificationStamp(document);
 				formatter.requestFormatting(document, textSelection).thenAcceptAsync(
-						edits -> shell.getDisplay().asyncExec(() -> formatter.applyEdits(document, edits)));
+						edits -> shell.getDisplay().asyncExec(() -> {
```

Comment:
This looks like you need to rebase the PR, I already commit  version of it which does the same (and to my taste is nicer because it relies on an incremental document version and not timestamps).